### PR TITLE
[WIP] New Vector visual

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Trying my hand again at #3414 . This time, with a custom visual that has all the good things and none of the bad... The only thing is that this requires the `gl+` backend/context for geometry shaders; I don't know if that's a total dealbreaker... In any case, I though I'd put my attempt here for reference and discussion.

What's different from main:
- no need to generate and juggle meshes to display vectors
- the lines are always "facing the camera", so no weird effects due to the view direction
- allows for scaling and non-scaling width (like point sizes), so one can keep it fixed in screen space
    - potentially, we could do the same thing for length
- should be faster

Why not just `LineVisual`:
- it would force integer-only widths
- it only allows for setting screen-space width

https://user-images.githubusercontent.com/23482191/150786919-7b27e2bb-468e-48fc-a747-25205902c3fd.mp4

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by sofroniewn** [comment]
> The only thing is that this requires the gl+ backend/context for geometry shaders; I don't know if that's a total dealbreaker

I'm afraid I don't really know what the consequence of this will be. Does this mean things won't work for certain users? Or we become more "hardware" specific? Does @nclack have further insight here?

> no need to generate and juggle meshes to display vectors

This should be a good thing!

Can you maybe add some benchmark results to the comments in this PR. For example rendering 10_000 or more vectors with the old approach vs the new approach (or more!). I could imagine that this approach will be much faster / more interactive, but it would be good to know.

> the lines are always "facing the camera", so no weird effects due to the view direction

We could fix this in our current implementation by making the meshed lines to be full tubes rather than the simple things they are now. that would mean more triangles though

> Why not just LineVisual:

Yes, I ran into those limitations the first time, so glad to have another approach with them fixed here.

I do wonder if this is very successful if we'd want to use these Lines in our Shapes layer for edges too. But that can come later

**Comment by nclack** [comment]
> > The only thing is that this requires the gl+ backend/context for geometry shaders; I don't know if that's a total dealbreaker
> 
> I'm afraid I don't really know what the consequence of this will be. Does this mean things won't work for certain users? Or we become more "hardware" specific? Does @nclack have further insight here?

They should work for the vast majority of folks.

[Geometry shaders](https://www.khronos.org/opengl/wiki/Geometry_Shader) have been in core opengl since 3.2 (promoted from [GL_ARB_geometry_shader4](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_geometry_shader4.txt) 2009) and GL ES since 3.1 (promoted from [GL_EXT_geometry_shader](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_geometry_shader.txt)).  They require glsl 1.5 or higher. 

You should be able to check for extension support similar to what was done in #3990.

@sofroniewn What do you think about setting a minimum opengl support version to 3.2? Vispy's [seems](https://vispy.org/installation.html#:~:text=Hardware%20requirements,-VisPy%20makes%20heavy&text=VisPy%20thus%20requires%20a%20fairly,opposed%20to%20the%20fixed%20pipeline) to be 2.1 (2006), but 3.2 is a significant improvement (see the [opengl version timeline](https://www.khronos.org/opengl/wiki/History_of_OpenGL) for details).  We'd still want to gracefully fail when a capability isn't present.

**Comment by tlambert03** [comment]
> @sofroniewn What do you think about setting a minimum opengl support version to 3.2?

this would, for example, preclude [this user](https://github.com/napari/napari/issues/3988) from using napari.
I have some awareness of that situation, they're using napari via nomachine on a university-hosted server that has good bandwidth access to our file-storage servers.  So, while the GPU itself is rather new and capable (A100), i'm not immediately sure how easy it is to get better opengl support there (@nclack, maybe you know whether there are limitations in that sort of environment?)

Anyway, I'm generally supportive of dropping support for legacy platforms.  So seems like something we should consider, but it would be good to know whether we're further hampering the cluster/server/vnc use case

**Comment by brisvag** [comment]
> Can you maybe add some benchmark results to the comments in this PR. For example rendering 10_000 or more vectors with the old approach vs the new approach (or more!). I could imagine that this approach will be much faster / more interactive, but it would be good to know.

Yes, I wanted to start the conversation first to see if the `gl+` thisn was a big issue. I'll try out some stuff soon!

> I do wonder if this is very successful if we'd want to use these Lines in our Shapes layer for edges too. But that can come later

I would definitely do that in that case. Also, it should be relatively easy to add antialiasing as well (although it would be a bit broken still due to vispy/vispy#2208.

**Comment by nclack** [comment]
> i'm not immediately sure how easy it is to get better opengl support there

@tlambert03 [EGL](https://www.khronos.org/egl) is really good and generally makes it possible. But clusters/multi-user environments are hard to maintain and that's usually why I've seen people stuck on opengl 2.1. 

> it would be good to know whether we're further hampering the cluster/server/vnc use case

It would be hard to argue that we weren't. I wish there a way to
* get pr's like this one in. It moves in the right direction.
* not completely lose the cluster users
* not have to do too much work on compatibility shims

I'm really not sure what direction is best. I keep talking myself in/out of sticking with GL 2.1.

**Comment by tlambert03** [comment]
> It would be hard to argue that we weren't. I wish there a way to

yeah, maybe not possible.  still not a dealbreaker.  Can't be all the things to all the use cases... You also mentioned a path for graceful degradation... what might that look like here?

**Comment by nclack** [comment]
> what might that look like here?

Something like: If the capability check fails, then we disable vector layers and issue a warning when someone tries to instance the layer.

**Comment by brisvag** [comment]
We can keep normal GL lines as fallback. This would allow us to keep everything identical all the way down to `VectorVisual`, and only there handle the (small) difference between the two. Obviously the downside is inconsistent behaviour (e.g: not being able to set canvas-space width), but it's probably better than simply disabling vectors.

**Comment by sofroniewn** [comment]
> We can keep normal GL lines as fallback. This would allow us to keep everything identical all the way down to VectorVisual, and only there handle the (small) difference between the two. Obviously the downside is inconsistent behaviour (e.g: not being able to set canvas-space width), but it's probably better than simply disabling vectors.

This approach would still let us drop my custom line -> mesh code right? I guess that's nice. We should just emit a warning when this is happening

**Comment by brisvag** [comment]
I added a proxy for `Line` so that now, depending on the availability of the backend, it chooses the correct visual. The switch works on my machine if triggered manually, but I don't know how to test if the if clause works as intended (can I "disable" the GL+ capabilities of my GPU?).

All the names and warning messages are completely tentative, but the functionality is all there!

File: napari/_vispy/visuals/vectors.py [hunk 1]
```diff
@@ -1,9 +1,50 @@
-from vispy.scene.visuals import Mesh
+import warnings
 
+from ..utils.gl import get_gl_extensions
 from .clipping_planes_mixin import ClippingPlanesMixin
 
+GLPLUS = 'geometry_shader' in get_gl_extensions()
+# GLPLUS = False
```
Comment (brisvag) [inline napari/_vispy/visuals/vectors.py:7]: Manually set `GLPLUS = False` to test locally.

BTW, It would be nice to have some way to test this on CI as well!

Comment (brisvag) [reply napari/_vispy/visuals/vectors.py:7]: And of course the CI is screaming because it does not have gl+ and all the warnings are errors :P

**Comment by brisvag** [comment]
Now that vispy 0.9.5 is out, this should be working. However, CI doesn't work well with `gl+`; there must be a way to fix it, cause vispy works fine. @djhoese, maybe you know where the relevant bit of code is in the vispy CI?

**Comment by djhoese** [comment]
I could be wrong but I don't think our single geometry shader example runs in CI. I think we skip it. I would make an issue or a PR (if you have ideas) on vispy and let's see if we can pull others into the conversation.

**Comment by brisvag** [comment]
Is this of any use? https://github.com/openrndr/setup-opengl

**Comment by djhoese** [comment]
Probably wouldn't hurt.

**Comment by brisvag** [comment]
A part from the opengl CI issues, I have one error that's a bit complicated. [This failure of test_perf.py](https://github.com/napari/napari/runs/5235616937?check_suite_focus=true#step:9:649) was apparently caused by the fact that [this function is being run](https://github.com/napari/napari/blob/e6ebaaee1005441a0ad193d01d9ca786cd1b70d9/napari/_vispy/utils/gl.py#L38-L42), which is needed in order to choose what `Line` visual to use. I guess this somehow spins up an `app`, and this somehow causes the failure? @tlambert03 you wrote ` get_gl_extensions`, maybe you have ideas?

**Comment by tlambert03** [comment]
> A part from the opengl CI issues, I have one error that's a bit complicated. [This failure of test_perf.py](https://github.com/napari/napari/runs/5235616937?check_suite_focus=true#step:9:649) was apparently caused by the fact that [this function is being run](https://github.com/napari/napari/blob/e6ebaaee1005441a0ad193d01d9ca786cd1b70d9/napari/_vispy/utils/gl.py#L38-L42), which is needed in order to choose what `Line` visual to use. I guess this somehow spins up an `app`, and this somehow causes the failure? @tlambert03 you wrote ` get_gl_extensions`, maybe you have ideas?

yeah, that function uses `_opengl_context()`, which creates a vispy `Canvas` (necessary to query gl parameters of the machine), and I think the default behavior of `Canvas` is to call `use_app(call_reuse=False)`.
If it's failing because `test_perf` needed to be the one to create the app first, then that feels more to me to be an idiosyncrasy of how one _must_ use perfmon.  I would consider just skipping that test (since it seems like a situation that will never happen if someone starts napari with the sole purpose of using perfmon)

**Comment by brisvag** [comment]
> I would consider just skipping that test (since it seems like a situation that will never happen if someone starts napari with the sole purpose of using perfmon)

But that would mean skipping the test always? Cause this check always needs to happen on import. Alternatively, is there a way to "close" the app once it's open?

**Comment by tlambert03** [comment]
> Alternatively, is there a way to "close" the app once it's open?

perfmon originally had something like this, where it would try to patch an existing  qt application by deleting and injecting a new one.  It was _super_ error prone and we got rid of it (https://github.com/napari/napari/pull/3171).

if you don't want to skip the test, another option is to perhaps just monkeypatch that one `get_gl_extensions` function into a no-op

**Comment by brisvag** [comment]
@tlambert03 and I have tried troubleshooting this GL+CI thing, to no avail. @nclack, I think we really need you for this!

Along the way, it turned out that macOS (also on Talley's laptop) has some compatibility complaints that result in `geometry_shader` being in the GL extensions, (and therefore correctly enabling the GL+ context), but then the `glsl` parsing fails because version 150 (a very old one that should definitely be supported) is deemed unsupported. Indeed, `glxinfo` (and here my understanding of the topic reaches its limits) seems to only claim support of glsl `1.2`.

So, the status is this:
- The new visual would be the default, so we probably want to have some level of CI testing of this and need to solve this problem one way or another
- The fallback visual (using simple opengl line primitives) is in some things worse than the current `main`, so we also should figure out and solve this mac issue, or we're throwing back a chunk of our users.

An alternative is to use the *current* behaviour as a fallback. The downside is that we don't get rid of the complexity of juggling meshes, but simply add to it, and things will be a bit of a pain to maintain. At that point, I'm not entirely sure if this PR is worth the additional maintenance effort!

**Comment by perlman** [comment]
After a nice (painful?) hacking session with @nclack and @brisvag, I found the following to work for getting OpenGL 4.1 & GLSL 4.10 on macOS:

```python
from PyQt5.QtGui import  QSurfaceFormat
from PyQt5.QtWidgets import QApplication, QOpenGLWidget

from PyQt5.QtWidgets import QApplication, QOpenGLWidget
from OpenGL import GL as gl

def main():
    # From https://forum.qt.io/topic/63581/qopenglwidget-5-5-opengl-4-1-macos-x-intel-hd-graphics-5000-1536-mb/4
    format = QSurfaceFormat()
    format.setVersion(4,1)
    format.setProfile(QSurfaceFormat.CoreProfile)
    QSurfaceFormat.setDefaultFormat(format)

    app = QApplication([])
    widget = QOpenGLWidget()
    widget.show()

    info = """
        Vendor: {0}
        Renderer: {1}
        OpenGL Version: {2}
        Shader Version: {3}
    """.format(
        gl.glGetString(gl.GL_VENDOR),
        gl.glGetString(gl.GL_RENDERER),
        gl.glGetString(gl.GL_VERSION),
        gl.glGetString(gl.GL_SHADING_LANGUAGE_VERSION)
    )

    print(info)

main()
```


Output on my laptop (macOS 12.2.1):
```
        Vendor: b'ATI Technologies Inc.'
        Renderer: b'AMD Radeon Pro 5600M OpenGL Engine'
        OpenGL Version: b'4.1 ATI-4.7.103'
        Shader Version: b'4.10'
```

But, napari (vispy?) isn't the happiest. We can dig in more next week...
image

**Comment by tlambert03** [comment]
> But, napari (vispy?) isn't the happiest. We can dig in more next week...

I've seen that before!  I think there's even an issue with a similar screenshot somewhere.  if only I could remember where! :joy:

**Comment by brisvag** [comment]
I think it would be helpful to get some additional input: @almarklein, `git blame` says you wrote a lot of this low level code in vispy. Do you have any suggestions?

**Comment by perlman** [comment]
We should probably split off the "macOS OpenGL 3.2/4.1" issue into it's own thread.

I suspect vispy will be the place to address this -- though it might be PyQT or QT itself. (I found references to the macOS OpenGL version oddities in the QT source code as well.)

**Comment by almarklein** [comment]
> I think it would be helpful to get some additional input: @almarklein, git blame says you wrote a lot of this low level code in vispy. Do you have any suggestions?

Could you summarize the problem you want input for? :)

**Comment by brisvag** [comment]
Sorry, I didn't realize how noodly the conversation became :P There are 2 main issues:

1. macos (both locallyt and on CI) does not initialize the right GL context. We are trying to dive in the guts of vispy/qt/openGL (this is what @perlman's comment above is about) to figure out how to force macos to give us a non-legacy version of opengl and glsl context. We have a hard time figuring out exactly where the first context is created in vispy, and how to hijack it. We think [it happens here (at least for qt5, non-egl backend)](https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/app/backends/_qt.py#L786) and somwhere in that class ([maybe here](https://github.com/vispy/vispy/blob/1df73c0d3683f806ed052c36a51ecb3b66382c86/vispy/app/backends/_qt.py#L879-L882)) is where we need to do something. Are we on the right track? Are you familiar with the opengl macos "quirks"?

2. We would like to test a visual with geometry-shaders in the CI. So far, I wasn't able to get the CI to give me the `gl+` context to do so, and I'm not even sure it's possible.

**Comment by almarklein** [comment]
> Are we on the right track? Are you familiar with the opengl macos "quirks"?

I don't know much about the macos quirks. It looks like you can pass e.g. [glformat](https://doc.qt.io/qt-5/qglformat.html) arg to `WGLWidget.__init__`, so that might be the way to go? 

The `initializeGL` is a hook, called by Qt when GL is initialized (whatever that means).

> We would like to test a visual with geometry-shaders in the CI [...] I'm not even sure it's possible.

Me neither :/

**Comment by brisvag** [comment]
> But, napari (vispy?) isn't the happiest. We can dig in more next week... image

Wait, what am I looking at? Is this just a screenshot of a messed up napari, and you're using napari to show it?

**Comment by sofroniewn** [comment]
Trying to understand what's going on in this post ;-)

image

**Comment by perlman** [comment]
> Wait, what am I looking at? Is this just a screenshot of a messed up napari, and you're using napari to show it?

You are looking at what happens with napari when OpenGL 4.1 is enabled on macOS.  (That is, set the default shader to the magic shader which enables 4.1 compatibility, and then start the napari viewer.)

The other sign that something is wrong are the error messages:
`WARNING: Could not create NSOpenGLContext with shared context, falling back to unshared context.`

**Comment by brisvag** [comment]
So somehow the vispy *canvas* got replaced with a flipped version of the whole napari *viewer*? This is black magic xD

**Comment by tlambert03** [comment]
I swear there's another issue about this from a year or two ago.  I have a vague recollection of even figuring it out.  Maybe search issues for that shared context warning?

**Comment by tlambert03** [comment]
Ah here it is: https://github.com/napari/napari/issues/2584

No grand solution, but maybe more clues

**Comment by perlman** [comment]
> Ah here it is: #2584
> 
> No grand solution, but maybe more clues

This is quite helpful -- ultimately, we need to figure out why the new NSOpenGLContext can't be created. The shared context is pretty funny here, with the window itself being the texture drawn within itself. :)

**Comment by Carreau** [comment]
This PR have not been active for quite some time. 

In order to keep the list of PRs short and easier to work with, I am going to tag this PR as [stale](https://github.com/napari/napari/pulls?q=is%3Apr+label%3Astale+is%3Aclosed) and close it.

Feel free to reopen and remove the "stale" label when work on this PR resumes. 
If you do not have the permissions to do so, feel free to request help from one of the maintainer to reopen and remove the `stale` label .
